### PR TITLE
nf48: Add firmware and docs.

### PR DESCRIPTION
**Pull request by haydenroche5**

### CODE REVIEW

**Comment by haydenroche5** [comment]
The GravityTDS library is unfortunately not available via the PlatformIO library registry, so I had to include it manually here.

File: 48-water-quality-monitor/README.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+| Sensor |  Docs | Calibration solution(s)
+| -------| ------| -----------------------
+| ORP    | [Link](https://wiki.dfrobot.com/Analog_ORP_Meter_SKU_SEN0165_) | [256 mV](https://www.amazon.com/calibration-solution-Oxidation-Reduction-Potential/dp/B01FS0G8WK)
+| pH     | [Link](https://wiki.dfrobot.com/Gravity__Analog_pH_Sensor_Meter_Kit_V2_SKU_SEN0161-V2) | [4.00](https://atlas-scientific.com/calibration-solutions/ph-4-00-calibration-solution-chem-ph-4/) and [7.00](https://atlas-scientific.com/calibration-solutions/ph-7-00-calibration-solution/)
+| TDS    | [Link](https://wiki.dfrobot.com/Gravity__Analog_TDS_Sensor___Meter_For_Arduino_SKU__SEN0244) | [1000 ppm](https://www.amazon.com/HM-Digital-1000ppm-Calibration-Solution/dp/B001FD6WZ8)
+
+The ORP calibration procedure involves computing an offset value. Once you have that value, edit `firmware/src/main.cpp` and replace 0 in the line `#define ORP_OFFSET 0` with your value.
```
[the remark this answers is not in the document]
Comment (haydenroche5) [reply 48-water-quality-monitor/README.md:59]: Sure, I can change this to an env var.

File: 48-water-quality-monitor/firmware/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+#endif
+
+#define ORP_SAMPLE_COUNT 25
+#define ORP_SAMPLE_DELAY_MS (1000 / ORP_SAMPLE_COUNT)
+
+#define PH_SAMPLE_COUNT 25
+#define PH_SAMPLE_DELAY_MS (1000 / PH_SAMPLE_COUNT)
+
+#define MAX_TEMP_SAMPLE_ATTEMPTS 3
+#define TEMP_SAMPLING_DELAY_MS 702
```
[the remark this answers is not in the document]
Comment (haydenroche5) [reply 48-water-quality-monitor/firmware/src/main.cpp:49]: Woops, this is supposed  to be 200. Must have made a typo. It comes from Zak's old project here: https://github.com/zfields/StreamResearch/blob/main/StreamResearch.ino#L27

File: 48-water-quality-monitor/firmware/src/main.cpp [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+void loop()
+{
+    // Fetch environment variables, if ENV_FETCH_INTERVAL_MS milliseconds have
+    // passed since the last fetch.
+    if (millis() - lastEnvFetchMs > ENV_FETCH_INTERVAL_MS) {
+        Serial.println("Fetching environment variables...");
+        lastEnvFetchMs = millis();
+
+        if (NotecardEnvVarManager_fetch(envVarManager, envVars, numEnvVars)
```
[the remark this answers is not in the document]
Comment (haydenroche5) [reply 48-water-quality-monitor/firmware/src/main.cpp:307]: It's whatever the default value is, which I don't know. But yeah, we should change it to something specific and document it in the README.

File: 48-water-quality-monitor/firmware/src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+
+    if (arr[0]<arr[1]) {
+        min = arr[0];
+        max = arr[1];
+    }
+    else {
+        min = arr[1];
+        max = arr[0];
+    }
+    for (i = 2; i < len; ++i) {
```
[the remark this answers is not in the document]
Comment (haydenroche5) [reply 48-water-quality-monitor/firmware/src/main.cpp:184]: It's weird. I copied this from DFRobot's sample code. I think I should just extract the averaging from `getPhVoltage` into another function and use that in both `getPhVoltage` and `getOrp`.

File: 48-water-quality-monitor/firmware/src/main.cpp [hunk 5]
```diff
@@ -319,30 +299,45 @@ void loop()
...
+        // sampleDs18b20 returning NAN), then we don't add the temp, ph, and
+        // tds fields, which all depend on this value.
+        if (!isnan(temp)) {
+            JAddNumberToObject(body, "temp", temp);
+            Serial.print("Temperature: ");
+            Serial.println(temp);
+
+            float phVoltage = sampleAverageAnalogSignal(PH_PIN) / ADC_RANGE *
+                              REFERENCE_MV;
+            float phValue = ph.readPH(phVoltage, temp);
```
Comment (m-mcgowan) [inline 48-water-quality-monitor/firmware/src/main.cpp:324]: Nice to see the pH sensor compensates for temperature

**Review by m-mcgowan — APPROVED** [review]

**Pull request merged**